Sources/NIOHPACK/HPACKEncoder.swift

```diff
@@ -0,0 +1,210 @@
+    /// This is defined as the sum of [name] + [value] + 32 for each header.
+    public var dynamicTableSize: Int {
+        return self.headerIndexTable.dynamicTableLength
+    }
+    
+    /// The maximum allowed size to which the dynamic header table may grow.
+    public private(set) var maxDynamicTableSize: Int {
+        get { return self.headerIndexTable.maxDynamicTableLength }
+        set { self.headerIndexTable.maxDynamicTableLength = newValue }
+    }
+    
+    /// Sets the maximum size for the dynamic table and optionally encodes the new value
+    /// into the current packed header block to send to the peer.
+    ///
+    /// - Parameter size: The new maximum size for the dynamic header table.
+    public mutating func setMaxDynamicTableSize(_ size: Int) {
+        self.maxDynamicTableSize = size
+        self.buffer.write(encodedInteger: UInt(size), prefix: 5, prefixBits: 0x20)
+    }
+    
+    /// Initializer and returns a new HPACK encoder.
+    ///
+    /// - Parameters:
+    ///   - allocator: An allocator for `ByteBuffer`s.
+    ///   - maxDynamicTableSize: An initial maximum size for the encoder's dynamic header table.
+    public init(allocator: ByteBufferAllocator, useHuffmanEncoding: Bool = true, maxDynamicTableSize: Int = HPACKEncoder.defaultDynamicTableSize) {
+        self.headerIndexTable = IndexedHeaderTable(allocator: allocator, maxDynamicTableSize: maxDynamicTableSize)
+        self.buffer = allocator.buffer(capacity: HPACKEncoder.defaultDataBufferSize)
+        self.useHuffmanEncoding = useHuffmanEncoding
+    }
+    
+    /// Reset the internal buffer, ready to begin encoding a new header block.
+    public mutating func reset() {
+        buffer.clear()
+    }
+    
+    /// Appends headers in the default fashion: indexed if possible, literal+indexable if not.
+    ///
+    /// - Parameter headers: A sequence of key/value pairs representing HTTP headers.
+    public mutating func append<S: Sequence>(headers: S) throws where S.Element == (name: String, value: String) {
+        try self.append(headers: headers.lazy.map { HeaderDefinition(name: $0.0, value: $0.1, indexing: .indexable) })
+    }
+    
+    /// Appends headers with their specified indexability.
+    ///
+    /// - Parameter headers: A sequence of key/value/indexability tuples representing HTTP headers.
+    public mutating func append<S : Sequence>(headers: S) throws where S.Element == HeaderDefinition {
+        for header in headers {
+            switch header.indexing {
+            case .indexable:
+                try self._appendIndexed(header: header.name.utf8, value: header.value.utf8)
+            case .nonIndexable:
+                self._appendNonIndexed(header: header.name.utf8, value: header.value.utf8)
+            case .neverIndexed:
+                self._appendNeverIndexed(header: header.name.utf8, value: header.value.utf8)
+            }
+        }
+    }
+    
+    public mutating func append(headers: HPACKHeaders) throws {
```

Comment:
D'oh, I missed the bit about sending both the lowest and highest. Seems obvious in retrospect. This looks like a good plan to me, though I'd probably rename `reset()` and `finalize()` as something that more clearly states their intent, like `beginEncoding(allocator:)`, taking a `ByteBufferAllocator`, and `endEncoding() -> ByteBuffer`. That way they pass in an allocator at the start and obtain the buffer at the end, and that's the only way to create or obtain the encoding buffer.

Does that sound ok?

Reply:
It would be *marvellous* if I could have the `encoding` state hold on to a `ByteBuffer`. Alas, it'll only cause a whole bunch of memory copies as it gets pulled out, found to be non-unique when mutated, and then put back in. Sigh…